plugins/logs/plugin.go

```diff
@@ -415,21 +415,42 @@ func (c *Config) validateAndInjectDefaults(services []string, pluginsList []stri
 
 // Plugin implements decision log buffering and uploading.
 type Plugin struct {
-	manager  *plugins.Manager
-	config   Config
-	buffer   *logBuffer
-	enc      *chunkEncoder
-	mtx      sync.Mutex
-	stop     chan chan struct{}
-	reconfig chan reconfigure
-	mask     *rego.PreparedEvalQuery
-	maskOnce *sync.Once
-	drop     *rego.PreparedEvalQuery
-	dropOnce *sync.Once
-	limiter  *rate.Limiter
-	metrics  metrics.Metrics
-	logger   logging.Logger
-	status   *lstat.Status
+	manager      *plugins.Manager
+	config       Config
+	buffer       *logBuffer
+	enc          *chunkEncoder
+	mtx          sync.Mutex
+	stop         chan chan struct{}
+	reconfig     chan reconfigure
+	preparedMask prepareOnce
+	preparedDrop prepareOnce
+	limiter      *rate.Limiter
+	metrics      metrics.Metrics
+	logger       logging.Logger
+	status       *lstat.Status
+}
+
+type prepareOnce struct {
+	once          *sync.Once
+	preparedQuery *rego.PreparedEvalQuery
+	err           error
+}
+
+func newPrepareOnce() *prepareOnce {
+	return &prepareOnce{
+		once: new(sync.Once),
+	}
+}
+
+func (po *prepareOnce) drop() {
+	po.once = new(sync.Once)
+}
+
+func (po *prepareOnce) prepare(f func() (*rego.PreparedEvalQuery, error)) (*rego.PreparedEvalQuery, error) {
```

Comment:
Hmm this is used both for preparing and for returning the previously prepared query state 🤔 So I think the name could be a bit more on-the-nose...? Maybe `cachedPrepared`? `prepareOnce`?

Reply:
I actually first had it as `prepareOnce()`, but changed it to not stutter 😄. But since it won't be a stuttering on the call-site, I can change it back for clarity 👍.

Reply:
Thanks!